### PR TITLE
Synthetic turbulence injection implementation 

### PR DESCRIPTION
**Pull request by gantech**
Most of the credit goes to @sayerhs. This is just a copy from https://github.com/sayerhs/nalu-wind/tree/f/synthturb with adaptations to change to amr-wind coding style and other small utilities available.

### CODE REVIEW

File: amr-wind/physics/SyntheticTurbulence.H [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  vs::Vector origin;
+
+  // Transformation matrix to convert from global coordinate system to local
+  // coordinate system
+  vs::Tensor tr_mat;
+
+  // Perturbation velocities (2, ny, nz)
+  std::vector<double> uvel;
+  std::vector<double> vvel;
+  std::vector<double> wvel;
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.H:61]:
```suggestion
  amrex::Vector<double> uvel;
  amrex::Vector<double> vvel;
  amrex::Vector<double> wvel;
```

Just to be consistent with rest of the code

File: amr-wind/physics/SyntheticTurbulence.H [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+class SyntheticTurbulence: public Physics::Register<SyntheticTurbulence>
+{
+public:
+
+  static const std::string identifier() { return "SyntheticTurbulence"; }
+
+  struct NCBoxTurb
+  {
+    std::string filename;
+  };
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.H:91]: Not needed anymore?

File: amr-wind/physics/SyntheticTurbulence.H [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+  SyntheticTurbulence(const CFDSim& sim);
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.H:93]:
```suggestion
  explicit SyntheticTurbulence(const CFDSim& sim);
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 4]
```diff
@@ -0,0 +1,574 @@
+#include "amr-wind/physics/SyntheticTurbulence.H"
+#include "amr-wind/CFDSim.H"
+#include "AMReX_iMultiFab.H"
+#include "AMReX_MultiFabUtil.H"
+#include "AMReX_ParmParse.H"
+#include "amr-wind/utilities/ncutils/nc_interface.H"
+#include "amr-wind/utilities/trig_ops.H"
+#include "amr-wind/utilities/tensor_ops.H"
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:8]:
```suggestion
#include "amr-wind/physics/SyntheticTurbulence.H"
#include "amr-wind/CFDSim.H"
#include "amr-wind/utilities/ncutils/nc_interface.H"
#include "amr-wind/utilities/trig_ops.H"
#include "amr-wind/utilities/tensor_ops.H"

#include "AMReX_iMultiFab.H"
#include "AMReX_MultiFabUtil.H"
#include "AMReX_ParmParse.H"
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+  dx.get(turb_grid.dx.data());
+
+  ncf.close();
+
+  // Create data structures to store the perturbation velocities for two planes
+  const size_t gridSize = 2 * ny * nz;
+  turb_grid.uvel.resize(gridSize);
+  turb_grid.vvel.resize(gridSize);
+  turb_grid.wvel.resize(gridSize);
+}
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:121]:
```suggestion
void process_nc_file(
  std::string& turb_filename,
  SynthTurbData& turb_grid)
{
#ifdef AMR_WIND_USE_NETCDF
  auto ncf = ncutils::NCFile::open(turb_filename, NC_NOWRITE);

  // Grid dimensions
  AMREX_ASSERT(ncf.dim("ndim").len() == AMREX_SPACEDIM);
  auto nx = ncf.dim("nx").len();
  auto ny = ncf.dim("ny").len();
  auto nz = ncf.dim("nz").len();

  turb_grid.box_dims[0] = nx;
  turb_grid.box_dims[1] = ny;
  turb_grid.box_dims[2] = nz;

  // Box lengths and resolution
  auto box_len = ncf.var("box_lengths");
  box_len.get(turb_grid.box_len.data());
  auto dx = ncf.var("dx");
  dx.get(turb_grid.dx.data());

  ncf.close();

  // Create data structures to store the perturbation velocities for two planes
  const size_t gridSize = 2 * ny * nz;
  turb_grid.uvel.resize(gridSize);
  turb_grid.vvel.resize(gridSize);
  turb_grid.wvel.resize(gridSize);
#else 
  amrex::ignore_unused(turb_filename, turb_grid);
#endif 
}
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+    uvel.get(&turb_grid.uvel[offset], start, count);
+    vvel.get(&turb_grid.vvel[offset], start, count);
+    wvel.get(&turb_grid.wvel[offset], start, count);
+  }
+
+  // Update left and right indices for future checks
+  turb_grid.ileft = il;
+  turb_grid.iright = ir;
+
+  ncf.close();
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:167]:
```suggestion
{
#ifdef AMR_WIND_USE_NETCDF
  auto ncf = ncutils::NCFile::open(turb_filename, NC_NOWRITE);

  std::vector<size_t> start{ {static_cast<size_t>(il), 0, 0} };
  std::vector<size_t> count{ {
          2, static_cast<size_t>(turb_grid.box_dims[1]),
          static_cast<size_t>(turb_grid.box_dims[2])} };

  auto uvel = ncf.var("uvel");
  auto vvel = ncf.var("vvel");
  auto wvel = ncf.var("wvel");

  if ((ir - il) == 1) {
    // two consequtive planes load them in one shot
    uvel.get(&turb_grid.uvel[0], start, count);
    vvel.get(&turb_grid.vvel[0], start, count);
    wvel.get(&turb_grid.wvel[0], start, count);
  } else {
    // Load the planes separately
    count[0] = 1;
    uvel.get(&turb_grid.uvel[0], start, count);
    vvel.get(&turb_grid.vvel[0], start, count);
    wvel.get(&turb_grid.wvel[0], start, count);

    start[0] = static_cast<size_t>(ir);
    const size_t offset = turb_grid.box_dims[1] * turb_grid.box_dims[2];
    uvel.get(&turb_grid.uvel[offset], start, count);
    vvel.get(&turb_grid.vvel[offset], start, count);
    wvel.get(&turb_grid.wvel[offset], start, count);
  }

  // Update left and right indices for future checks
  turb_grid.ileft = il;
  turb_grid.iright = ir;

  ncf.close();
#else 
  amrex::ignore_unused(turb_filename, turb_grid, il, ir);
#endif 
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+  const CFDSim& sim)
+    : m_time(sim.time()),
+      m_repo(sim.repo()),
+      m_mesh(sim.mesh()),
+      m_velocity(sim.repo().get_field("velocity")),
+      m_density(sim.repo().get_field("density")),
+      m_turb_force(sim.repo().declare_field("synth_turb_forcing", 3))
+{
+  const amrex::Real pi = std::acos(-1.0);
+
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:323]:
```suggestion
{
#ifndef AMR_WIND_USE_NETCDF
    amrex::Abort("SyntheticTurbulence: AMR-Wind was not built with NetCDF support.");
#endif
  const amrex::Real pi = std::acos(-1.0);

```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+  m_turb_grid.tr_mat.xx() = -std::sin(wind_direction);
+  m_turb_grid.tr_mat.xy() = -std::cos(wind_direction);
+  m_turb_grid.tr_mat.xz() = 0.0;
+  // z always points upwards (for now...)
+  m_turb_grid.tr_mat.zx() = 0.0;
+  m_turb_grid.tr_mat.zy() = 0.0;
+  m_turb_grid.tr_mat.zz() = 1.0;
+  // y = z .cross. x
+  m_turb_grid.tr_mat.y() = m_turb_grid.tr_mat.z() ^ m_turb_grid.tr_mat.x();
+  //utils::cross_prod(&m_turb_grid.tr_mat[2][0], &m_turb_grid.tr_mat[0][0], &m_turb_grid.tr_mat[1][0]);
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:427]:
```suggestion
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+
+void SyntheticTurbulence::initialize_fields(
+    int // level
+    ,
+    const amrex::Geometry& // geom)
+)
+{
+    //TODO: Figure out what goes here
+}
+
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:453]:
```suggestion
void SyntheticTurbulence::initialize_fields(int, const amrex::Geometry&)
{}
```

File: amr-wind/physics/SyntheticTurbulence.H [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  void post_regrid_actions() override {}
+
+  void pre_advance_work() override;
+
+  void post_advance_work() override {}
+
+private:
+  void initialize();
+
+  void update();
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.H:114]: Needs to be public for CUDA

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+  const amrex::Real curTime = m_time.new_time() - m_time_offset;
+  const amrex::Real eqivLen = m_wind_profile->reference_velocity() * curTime;
+  int il, ir;
+  get_lr_indices(m_turb_grid, 0, eqivLen, il, ir);
+  load_turb_plane_data(m_turb_filename, m_turb_grid, il, ir);
+
+  m_is_init = false;
+}
+
+void SyntheticTurbulence::update()
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:475]: Make this member public in header

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+  const int sdir = (*m_wind_profile).shear_dir();
+
+  const int nlevels = repo.num_active_levels();
+  for (int lev=0; lev < nlevels; ++lev) {
+      const auto& geom = geom_vec[lev];
+      const auto& problo = geom.ProbLoArray();
+
+      const amrex::Real dx = geom.CellSize()[0];
+      const amrex::Real dy = geom.CellSize()[1];
+      const amrex::Real dz = geom.CellSize()[2];
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:502]:
```suggestion
      const amrex::Real dx = geom.CellSize()[0];
      const amrex::Real dy = geom.CellSize()[1];
      const amrex::Real dz = geom.CellSize()[2];
      const auto& trmat = m_turb_grid.tr_mat;
      const auto& origin = m_turb_grid.origin;
```

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+            vs::Vector vel_g;
+
+            xyz_g[0] = problo[0] + (i + 0.5) * dx;
+            xyz_g[1] = problo[1] + (j + 0.5) * dy;
+            xyz_g[2] = problo[2] + (k + 0.5) * dz;
+
+            // Transform position vector from global inertial
+            // reference frame to local reference frame attached to
+            // the turbulence grid.
+            xyz_l = m_turb_grid.tr_mat & ( xyz_g - m_turb_grid.origin);
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:526]:
```suggestion
            xyz_l = trmat & ( xyz_g - origin);
```

No `m_*` class instance variable can appear inside `amrex::ParallelFor`

File: amr-wind/physics/SyntheticTurbulence.H [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+  inline int shear_dir() const { return m_shear_dir; }
+
+protected:
+  const double m_ref_vel;
+  const int m_shear_dir;
+};
+
+} // namespace synth_turb
+
+struct SynthTurbData
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.H:40]: We need this to be safe to go on device.

File: amr-wind/physics/SyntheticTurbulence.cpp [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+
+            // Check if the point is in the box, if not we skip this
+            // node. The function will also populate the interpolation
+            // weights for points that are determined to be within the
+            // box.
+            bool ptInBox = find_point_in_box(m_turb_grid, xyz_l, wts_loc);
+            if (ptInBox) {
+                // Interpolate perturbation velocities in the local
+                // reference frame
+                interp_perturb_vel(m_turb_grid, wts_loc, vel_l);
```
Comment (sayerhs) [inline amr-wind/physics/SyntheticTurbulence.cpp:538]: These need to be updated for CUDA.

File: amr-wind/physics/SyntheticTurbulence.cpp (hunk 11 above)
Comment (gantech) [reply amr-wind/physics/SyntheticTurbulence.cpp:477]: Fixed in https://github.com/Exawind/amr-wind/pull/327/commits/5213aab7053bb736e11a39c02ff753c2820f8837

File: amr-wind/physics/SyntheticTurbulence.H (hunk 10 above)
Comment (gantech) [reply amr-wind/physics/SyntheticTurbulence.H:108]: Fixed in https://github.com/Exawind/amr-wind/pull/327/commits/5213aab7053bb736e11a39c02ff753c2820f8837

File: amr-wind/physics/SyntheticTurbulence.H (hunk 2 above)
Comment (gantech) [reply amr-wind/physics/SyntheticTurbulence.H]: Fixed in https://github.com/Exawind/amr-wind/pull/327/commits/5213aab7053bb736e11a39c02ff753c2820f8837

**Pull request merged**